### PR TITLE
Support `radial-gradient` (1/2)

### PR DESCRIPTION
**Pull request by You-J**
Supports `radial gradient` with two or more colors. Only colors are supported without additional properties such as center or radius.

in figma
Screen Shot 2021-11-11 at 7 44 27 PM

in editor
Screen Shot 2021-11-11 at 7 40 08 PM

demo1: http://localhost:6626/to-code?framework=react&component=false&preview=vanilla&design=https://www.figma.com/file/iypAHagtcSp3Osfo2a7EDz/engine?node-id=4105%3A3903

demo1- flutter: http://localhost:6626/to-code?framework=flutter&component=false&preview=vanilla&design=https://www.figma.com/file/iypAHagtcSp3Osfo2a7EDz/engine?node-id=4105%3A3903

demo2: http://localhost:6626/to-code?framework=react&component=false&preview=vanilla&design=https://www.figma.com/file/iypAHagtcSp3Osfo2a7EDz/engine?node-id=2659%3A3490

test link: http://localhost:6626/to-code?framework=react&component=false&preview=vanilla&design=https://www.figma.com/file/iypAHagtcSp3Osfo2a7EDz/engine?node-id=2684%3A4233

test link2: http://localhost:6626/to-code?framework=react&component=false&preview=vanilla&design=https://www.figma.com/file/iypAHagtcSp3Osfo2a7EDz/engine?node-id=2684%3A4515

# Figma Gradient

- [figma paint - gradient paint](https://www.figma.com/plugin-docs/api/Paint/#gradientpaint)
- [css gradient](https://developer.mozilla.org/en-US/docs/Web/CSS/gradient)

## gradient type

- linear gradient
- radial gradient
- angular gradient
- diamond gradient

### linear gradient

**css**

- [linear-gradient()](<https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/linear-gradient()>)

`linear-gradient()` css function

```css
background: linear-gradient(45deg, blue, red);
```

> **syntax**
>
> wip
**flutter**

- [LinearGradient()](https://api.flutter.dev/flutter/painting/LinearGradient-class.html)

### radial gradient

**css**

- [radial-gradient()](<https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/radial-gradient()>)

```css
background: radial-gradient(red, blue);
```

**flutter**

- [RadialGradient()](https://api.flutter.dev/flutter/painting/RadialGradient-class.html)

> **gradient paint opacity**
>
> gradient color's opacity \* fill's opacity
> **gradient paint degree**
> WIP

### CODE REVIEW

File: packages/builder-css-styles/radial-gradient/index.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { color } from "../color";
+
+/**
+ * https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/radial-gradient()
+ *
+ * @param g
+ * @returns
+ */
+export function radialGradient(g: RadialGradientManifest): CSSProperty.Color {
+  // throw "css gradient not ready";
```
Comment (softmarshmallow) [inline packages/builder-css-styles/radial-gradient/index.ts:12]: rm line

File: packages/designto-flutter/painting/painting-box-decoration.ts [hunk 2]
```diff
@@ -20,6 +22,7 @@ function fromBackground(b: Background): BoxDecoration {
     switch (b.type) {
       case "gradient": {
         console.error("gradient bg not ready");
```
Comment (softmarshmallow) [inline packages/designto-flutter/painting/painting-box-decoration.ts:24]: Was gradient on flutter never been supported? if so remove the `console.log`

File: packages/designto-flutter/painting/painting-box-decoration.ts [hunk 3]
```diff
@@ -33,8 +36,25 @@ function fromBackground(b: Background): BoxDecoration {
...
+    case GradientType.RADIAL: {
+      return new BoxDecoration({
+        gradient: radialGradient(g),
+      });
+    }
+    default: {
+      // TODO: add;
+      // GRADIENT_ANGULAR;
+      // GRADIENT_DIAMOND;
+      throw "not ready.";
```
Comment (softmarshmallow) [inline packages/designto-flutter/painting/painting-box-decoration.ts:55]:
```ts
throw new Error(`Gradient type of "${g}" is not yet supported on flutter platform.`)
```

File: packages/designto-flutter/painting/painting-box-decoration.ts [hunk 4]
```diff
@@ -33,8 +36,25 @@ function fromBackground(b: Background): BoxDecoration {
   }
 }
 
-function fromGradient(): BoxDecoration {
-  throw "not ready.";
+function fromGradient(g: Gradient): BoxDecoration {
+  switch (g._type) {
```
Comment (softmarshmallow) [inline packages/designto-flutter/painting/painting-box-decoration.ts:40]: will it not be more safer to use `g?._type`

File: packages/designto-flutter/painting/painting-box-decoration.ts [hunk 5]
```diff
@@ -33,8 +36,25 @@ function fromBackground(b: Background): BoxDecoration {
...
+      return new BoxDecoration({
+        gradient: linearGradient(g),
+      });
+    }
+    case GradientType.RADIAL: {
+      return new BoxDecoration({
+        gradient: radialGradient(g),
+      });
+    }
+    default: {
```
Comment (softmarshmallow) [inline packages/designto-flutter/painting/painting-box-decoration.ts:51]: + add case
```
case undefined: {
   return;
}
```

(Just a suggestion)

File: packages/designto-token/token-gradient/gradient.ts [hunk 6]
```diff
@@ -45,16 +48,22 @@ export function tokenize_gradient(
...
       });
     case "GRADIENT_RADIAL":
+      return new RadialGradient({
+        center: direction.begin,
+        colors: colors,
+        stops: stopPoints,
+        // TODO: support radius
+      });
+
       console.error("GRADIENT_RADIAL not handled");
```
Comment (softmarshmallow) [inline packages/designto-token/token-gradient/gradient.ts:58]: rm line

**Review by softmarshmallow — CHANGES_REQUESTED** [review]
Better than i expected. LGTM overall. few fix required

File: packages/designto-flutter/painting/painting-box-decoration.ts (hunk 2 above)
Comment (You-J) [reply packages/designto-flutter/painting/painting-box-decoration.ts]: removed!

File: packages/builder-css-styles/radial-gradient/index.ts (hunk 1 above)
Comment (You-J) [reply packages/builder-css-styles/radial-gradient/index.ts]: removed

File: packages/designto-token/token-gradient/gradient.ts (hunk 6 above)
Comment (You-J) [reply packages/designto-token/token-gradient/gradient.ts]: removed!

File: packages/designto-flutter/painting/painting-box-decoration.ts (hunk 5 above)
Comment (You-J) [reply packages/designto-flutter/painting/painting-box-decoration.ts:55]: I think there could be a contingency. Added with comments.

File: packages/designto-flutter/painting/painting-box-decoration.ts (hunk 4 above)
Comment (You-J) [reply packages/designto-flutter/painting/painting-box-decoration.ts]: fixed!

**Pull request merged**